index.bs

```diff
@@ -931,46 +931,67 @@ that can be written to, to transmit data to the remote host.
 <pre class="idl">
 [ Exposed=(Window,Worker) ]
 interface SendStream : WritableStream /* of Uint8Array */ {
-  readonly attribute Promise&lt;StreamAbortInfo&gt; writingAborted;
-  undefined abortWriting(optional StreamAbortInfo abortInfo = {});
+  readonly attribute Promise&lt;StreamAbortInfo&gt; remoteReset;
+  undefined reset(optional StreamAbortInfo abortInfo = {});
 };
 </pre>
 
-## Overview ##  {#outgoing-stream-overview}
+## Overview ##  {#send-stream-overview}
 
 The {{SendStream}} will be initialized by running the {{WritableStream}}
 initialization steps.
 
-## Attributes ##  {#outgoing-stream-attributes}
+## Attributes ##  {#send-stream-attributes}
 
-: <dfn attribute for="SendStream">writingAborted</dfn>
-:: The `writingAborted` attribute represents a promise that is [=fulfilled=]
-    when the a message from the remote side aborting the stream is received.
-    For QUIC, that message is a STOP_SENDING frame. When the stream receives
-    this mesage, the user agent MUST run the following:
-      1. Let |stream| be the {{SendStream}} object.
-      1. Let |transport| be the {{WebTransport}}, which the |stream| was created
-         from.
-      1. Remove the stream from the |transport|'s [=[[OutgoingStreams]]=].
-      1. [=Resolve=] the promise with the resulting {{StreamAbortInfo}} with
-         the {{StreamAbortInfo/errorCode}} set to the value from the aborting
-         message from the remote side.
+: <dfn attribute for="SendStream">remoteReset</dfn>
+:: The `remoteReset` attribute represents a promise that is [=fulfilled=]
+    when a message from the remote side aborting the stream is received.
+    For QUIC, that message is a STOP_SENDING frame.
 
-## Methods ##  {#outgoing-stream-methods}
+## Methods ##  {#send-stream-methods}
 
-: <dfn method for="SendStream">abortWriting()</dfn>
+: <dfn method for="SendStream">reset()</dfn>
 :: A hard shutdown of the {{SendStream}}. It may be called regardless of
    whether the {{SendStream}} was created by the local or remote peer. When
-   the `abortWriting` method is called, the user agent MUST run the following
+   the `reset` method is called, the user agent MUST run the following
    steps:
      1. Let |stream| be the {{SendStream}} object which is about to abort
         writing.
      1. Let |transport| be the {{WebTransport}}, which the |stream| was created
         from.
-     1. Remove the stream from the transport's [=[[OutgoingStreams]]=].
+     1. Remove |stream| from the transport's [=[[OutgoingStreams]]=].
      1. Let |abortInfo| be the first argument.
-     1. Start the closing procedure by sending a RST_STREAM frame with its
-        error code set to the value of |abortInfo.errorCode|.
+     1. Let |reason| be |abortInfo|.errorCode.
+     1. If it hasn't started already, start the closing procedure by sending a
```

Comment:
It is, but my understanding is that [aborting](https://streams.spec.whatwg.org/#writable-stream-abort) in the streams spec is not guaranteed to be immediate (it's not possible to abort a stream that's already been closed for instance), which is the reason for having `reset()`.

So we're starting the closing procedure opportunistically from both places to have it happen sooner rather than later.

Reply:
> ```
> 3. Return ! WritableStreamAbort(this, abortInfo).
> ```

@yutakahirano My understanding is the whole point of `reset()` is that it works immediately where `abort()` wouldn't (e.g. when `"closed"`), so we can't rely on [WritableStreamAbort](https://streams.spec.whatwg.org/#writable-stream-abort), because among other things it says: _"2. If state is "closed" or "errored", return a promise resolved with undefined."_

Regarding checking if the stream is locked, this rename-or-remove PR tries to preserve how [abortWriting](https://w3c.github.io/webtransport/#dom-sendstream-abortwriting) is spec'ed to work today, while making the implied impact on streams a bit more explicit. So let's open a new issue on that.

> Alternatively, we could unify abort and reset by adding one more field to the "reason" argument.

We'd still need `reset()` for the same reason that `abort()` won't let us reset closed streams (which still may be flushing data the way the streams spec is designed).

> ```
> 1. Send a STOP_SENDING frame with |reason|.
> ```

Whoops, that was meant to be RESET_STREAM. My bad. I think my mistake confused you. Let me fix it and we start over.

Also, `remoteReset` on the sender is confusing. It should be `remoteCancelled` (since the remote end is a ReceiveStream with a cancel method).

Reply:
> My understanding is the whole point of reset() is that it works immediately where abort() wouldn't (e.g. when "closed"), so we can't rely on WritableStreamAbort, because among other things it says: "2. If state is "closed" or "errored", return a promise resolved with undefined."

WritableStream's state becomes "closed" or "errored" when the closing or erroring operation starts running, not when it is queued. You can run the following code


```
ws = new WritableStream({
  write(chunk) {
    console.log(`write(${chunk})`);
    return new Promise(resolve => setTimeout(resolve, 5000));
  },
  close(reason) {
    console.log(`close(${reason})`);
    return new Promise(resolve => setTimeout(resolve, 5000));
  },
  abort(reason) {
    console.log(`abort(${reason})`);
    return new Promise(resolve => setTimeout(resolve, 5000));
  }
}); writer = ws.getWriter();

writer.write(1); writer.write(2); writer.write(3); writer.close(); writer.abort();
```

and you will see that abort operation runs before the second chunk arrvies.

Your concern is reasonable for the [inflight write request](https://streams.spec.whatwg.org/#writablestream-inflightwriterequest) - In the above example, the abort operation waits for the first write request to be completed. https://github.com/whatwg/streams/issues/1015 may be a solution for the problem.

Reply:
@yutakahirano makes sense.

Meeting:
* Merge PR for forward progress to not block on future nice things, but with the intent that we would eventually be able to remove these methods in favor of making `abort` in the streams spec work for us.

Reply:
Thank you!

I filed #260. Let's discuss this there.